libs/prelude/Prelude/Interfaces.idr

```diff
@@ -150,41 +150,69 @@ public export
 mapHom : Bifunctor f => (a -> b) -> f a a -> f b b
 mapHom f = bimap f f
 
+||| An applicative functor without `pure`.
+|||
+||| This allows us to get access to `(<*>)` and
+||| possibly `(>>=)` for data types, for which we cannot
+||| define `pure`, such as sorted maps or vectors
+||| (fixed length lists). In case of vectors, it is possible
+||| to implement `pure`, but only if we have access to the
+||| vector's length (see the implementations of
+||| `Apply` and `Applicative` for `Data.Vect` in base).
+|||
+||| In the case of `Pair`, we can get `(<*>)` and
+||| `(>>=)` for pairs whose first value is a `Semigroup`, while
+||| for `pure` we'd need a `Monoid`.
 public export
-interface Functor f => Applicative f where
-  constructor MkApplicative
-  pure : a -> f a
+interface Functor f => Apply f where
+  constructor MkApply
   (<*>) : f (a -> b) -> f a -> f b
 
 public export
-(<*) : Applicative f => f a -> f b -> f a
+(<*) : Apply f => f a -> f b -> f a
 a <* b = map const a <*> b
 
 public export
-(*>) : Applicative f => f a -> f b -> f b
+(*>) : Apply f => f a -> f b -> f b
 a *> b = map (const id) a <*> b
 
-%allow_overloads pure
 %allow_overloads (<*)
 %allow_overloads (*>)
 
+namespace Apply
+  ||| Composition of applicative functors is an applicative functor.
+  public export
+  [Compose] (Apply f, Apply g) => Apply (f . g)
+    using Functor.Compose where
+      fun <*> x = (<*>) <$> fun <*> x
+
+public export
+interface Apply f => Applicative f where
+  constructor MkApplicative
+  pure : a -> f a
+
+%allow_overloads pure
+
 namespace Applicative
   ||| Composition of applicative functors is an applicative functor.
   public export
   [Compose] (Applicative f, Applicative g) => Applicative (f . g)
-    using Functor.Compose where
+    using Apply.Compose where
       pure = pure . pure
-      fun <*> x = [| fun <*> x |]
 
 public export
-interface Applicative f => Alternative f where
+interface Apply f => Alt f where
+  constructor MkAlt
+  (<|>) : f a -> Lazy (f a) -> f a
+
+public export
+interface Applicative f => Alt f => Alternative f where
   constructor MkAlternative
   empty : f a
```

Comment:
I guess this design decision echoes with #1472: Why should alternative be an `Alt Applicative` with a point rather than a `Pointed Applicative` (`interface Pointed a where point : a`) with
a `Sum`?

Should we flatten the hierarchy entirely?

Reply:
I'm afraid I can't follow you here. Could you please quickly explain how this is related to #1472 and how you'd like to flatten the hierarchy?

Reply:
So, I had a look at how Purescript does it and added an intermediate interface `Plus` plus some doc strings (which I more or less copied from Purescript as well). Reference: [purescript-control](https://pursuit.purescript.org/packages/purescript-control/5.0.0).

Reply:
In addition, I somewhat reorganized `Prelude.Interfaces` to more cleanly group interfaces in the same hierarchy.

Reply:
This is the relevant quote:

> Preorder is transitive and reflexive, and Equivalence is a symmetric preorder. But how does a "tolerance relation", which is reflexive and transitive, fit into this scheme? Why does an equivalence relation have to be defined as a symmetric preorder instead of a transitive tolerance relation? Why bring a preorder into the definition of equivalence at all?

Why do we define an applicative as an apply with pure rather than lift (`interface Lift f where pure : a -> f a`) with an apply?

In #1472 we decided to have separate constraints for:
* reflexive
* symmetric
* transitive
* etc.
and define common notion such as equivalence as aliases to the set of constraints they represent.

```idris
interface (Reflexive r, Symmetric r, Transitive r) => Equivalence r where
  -- nothing
```

rather than

```idris
interface Preorder r => Equivalence r where
  symmetric : r a b -> r b a
```


I'm asking whether we should be doing the same here:

```idris
interface Functor f where
  map : (a -> b) -> f a -> f b

interface Lift f where
  pure : a -> f a

interface Apply f where
  (<*>) : f (a -> b) -> f a -> f b

interface Bind f where
  (>>=) : f a -> (a -> f b) -> f b

interface (Lift f, Apply f) => Applicative f where
interface (Applicative f, Bind f) => Monad f where
```

Reply:
I'll give this a try and will report back here if some unexpected problems occurred. One advantage I see already that for Monads, `(<*>)` can be implemented in terms of `(>>=)`, which is tedious if we have a strict hierarchy. With your suggestion, we could just implement `Functor` and `Bind` first.

One disadvantage would be, that we generate some intermediary interfaces without a proper set of laws. Not sure, what the purists think of this...

Reply:
So, here's what I found: We need `Functor` => `Bind`, otherwise we will not be able to choose whether to implement `join` or `>>=` or both, which I consider to be quite a drawback. Therefore, if we (almost) completely flatten the `Monad` hierarchy, here's what I suggest:

* `Functor f` : `map : (a -> b) -> f a -> f b`
* `Apply f` : `(<*>) : f (a -> b) -> f a -> f b`
* `Lift f` : `pure : a -> f a`
* `Functor f => Bind f` : `(>>=) : f a -> (a -> f b) -> f b`
* `Functor f => Apply f => Semiapplicative f`
* `Semiapplicative f => Lift f -> Applicative f`
* `Bind f => Apply f => Semimonad f`
* `Semimonad f => Lift f => Monad f`

The names `Semiapplicative` and `Semimonad` are taken from [this issue in the semigroupoids library](https://github.com/ekmett/semigroupoids/issues/26).

In addition, we'll have

* `Alt f` : `(<|>) : f a -> f a -> f a`
* `Alt f => Plus f` : `empty : f a`
* `Alt f => Applicative f => Alternative f`

I haven't inlcuded this new version here yet, since this is quite a bit of work and I first would like to here your thoughts.

Reply:
After these experiments, just to give my personal opinion on this: I'd like to stick with the hierarchy suggested in #1325. This is, how it is done in Purescript and in haskell's semigroupoids library, so people familiar with one of these will be comfortable immediately. If we flatten the hierarchy, people will have to write (dummy) implementations for three more interfaces: `Lifted`, `Semiapplicative`, and `Semimonad`. We could give default definitions for the latter two, but some of these (for instance the IO ones) are marked as to be inlined, and I'm not sure how this will be affected by default definitions.

Therefore, I think the hierarchy as it is right now (in this PR) is the way to go.